setuptools/build_meta.py

```diff
@@ -110,51 +95,88 @@ def _file_with_extension(directory, extension):
     return file
 
 
-def _open_setup_script(setup_script):
-    if not os.path.exists(setup_script):
-        # Supply a default setup.py
-        return io.StringIO(u"from setuptools import setup; setup()")
-
-    return getattr(tokenize, 'open', open)(setup_script)
-
-
 @contextlib.contextmanager
 def suppress_known_deprecation():
     with warnings.catch_warnings():
         warnings.filterwarnings('ignore', 'setup.py install is deprecated')
         yield
 
 
+@contextlib.contextmanager
+def _patch_distutils_core():
+    """Make sure distutils.core uses the latest enhancements"""
+    orig_exec = exec
+    if hasattr(distutils.core, "run_commands"):
+        yield  # do nothing, already using the improved version of distutils
+        return
+
+    def _exec(code, global_vars):
+        try:
+            fid, tmp = tempfile.mkstemp(suffix=f"{uuid4()}-setup.py", text=False)
+            os.close(fid)  # Ignore the low level API
+            with open(tmp, "wb") as f:
+                f.write(code)
+            with tokenize.open(tmp) as f:
+                code = f.read().replace(r'\r\n', r'\n')
+        finally:
+            os.remove(tmp)
+        orig_exec(code, {**global_vars, "__name__": "__main__"})
+
+    def _run_commands(dist):
+        try:
+            dist.run_commands()
+        except Exception as ex:
+            raise SystemExit("error:" + str(ex))
+
+    distutils.core.exec = _exec
+    distutils.core.run_commands = _run_commands
+    try:
+        yield
+    finally:
+        distutils.core.exec = orig_exec
+        del distutils.core.run_commands
+
+
 class _BuildMetaBackend(object):
 
     def _fix_config(self, config_settings):
         config_settings = config_settings or {}
         config_settings.setdefault('--global-option', [])
         return config_settings
 
-    def _get_build_requires(self, config_settings, requirements):
-        config_settings = self._fix_config(config_settings)
+    def _get_dist(self):
+        """Retrieve a distribution object already configured."""
 
-        sys.argv = sys.argv[:1] + ['egg_info'] + \
-            config_settings["--global-option"]
-        try:
-            with Distribution.patch():
-                self.run_setup()
-        except SetupRequirementsError as e:
-            requirements += e.specifiers
+        if os.path.exists(SETUP_SCRIPT) and os.stat(SETUP_SCRIPT).st_size > 0:
```

Comment:
Could you put this condition into a label-variable so that it's self-explanatory?
```suggestion
        non_empty_setup_py_present = os.path.exists(SETUP_SCRIPT) and os.stat(SETUP_SCRIPT).st_size > 0
        if non_empty_setup_py_present:
```